tokio/src/time/driver/entry.rs

```diff
@@ -1,361 +1,721 @@
+    pub(crate) fn poll_elapsed(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<(), super::Error>> {
+        let this = unsafe { self.get_unchecked_mut() };
 
+        // Most of the time we just can do a quick check of our state value without taking the lock.
         loop {
-            // In these two cases, there is no work to do when resetting the
-            // timer. If the `Entry` is in an error state, then it cannot be
-            // used anymore. If resetting the entry to the current value, then
-            // the reset is a noop.
-            if curr == ERROR || curr == when {
-                return;
+            let state = this.inner().state.get(Ordering::Relaxed);
+            match state {
+                EntryState::NotRegistered => {
+                    unsafe {
+                        let handle = this.handle();
+                        this.driver.add_entry(handle);
+                    }
+
+                    // Re-poll afterward - the driver will communicate
+                    // synchronous completion or error states by updating our
+                    // state. If there was no such error, we'll go into the
+                    // Registered branch below to set the waker.
+                }
+                EntryState::Registered => {
+                    // Implicit acquire fence.
+                    this.inner().waker.register_by_ref(cx.waker());
+
+                    // Need to recheck the state in case we raced with timer
+                    // firing. The above register_by_ref forms an acquire fence
+                    // matched with the corresponding take operation (as a
+                    // release fence) in fire().
+                    if this.inner().state.get(Ordering::Relaxed) == EntryState::Registered {
+                        break Poll::Pending;
+                    }
+                }
+                EntryState::FiringInProgress => {
+                    // Synchronize with the driver, once we finish the true
+                    // completed state should be visible.
+                    unsafe {
+                        this.driver
+                            .sync_timer(NonNull::new(this.inner.get()).unwrap())
+                    };
+
+                    // We should now be in a completed state, so loop back around
+                }
+                EntryState::Fired => {
+                    // We're probably done, but if we were reset, it's possible we
+                    // raced with being fired by the driver. If this happened, we
+                    // reset our state and try again.
+
+                    // To check, it's sufficient to see if cached_when and true_when
+                    // are in sync. It's safe to access cached_when here because
+                    // writing FiringInProgress with a release barrier happens after
+                    // the last sync_when on the driver.
```

Comment:
Where is the acquire that synchronizes with the `FiringInProgress`?

Reply:
This is actually synchronized via the AtomicWaker (comment corrected)